dali/operators/audio/resample.cc

```diff
@@ -0,0 +1,168 @@
+
+This parameter cannot be specified together with ``in_rate`` and ``out_rate`` or ``out_length``.)",
+    nullptr, true)
+  .AddOptionalArg<int64_t>("out_length", R"(The requested output length, in samples.
+
+The scaling factor is the ratio of this output length to the input length. This parameter
+cannot be specified together with ``in_rate``, ``out_rate`` or ``scale``.)",
+    nullptr, true)
+  .AddOptionalArg("quality", R"(Resampling quality, where 0 is the lowest, and 100 is
+the highest.
+
+0 gives 3 lobes of the sinc filter, 50 gives 16 lobes, and 100 gives 64 lobes.)",
+    50.0f, false)
+  .AddOptionalArg<DALIDataType>("dtype", R"(The ouput type.
+
+If not specified, the output type is the same as the input type. When the type is changed,
+the values are normalized to fill the dynamic range of the target data type. When converting
+floating point inputs to integer types, the values are assumed to be in -1..1 range. When converting
+between signed and unsigned types, 0 translates to half-range of the unsigned type.)",
+    nullptr, false);
+
+namespace audio {
+
+using kernels::InTensorCPU;
+using kernels::OutTensorCPU;
+
+class ResampleCPU : public ResampleBase<CPUBackend> {
+ public:
+  using Base = ResampleBase<CPUBackend>;
+  explicit ResampleCPU(const OpSpec &spec) : Base(spec) {
+    auto params = ResamplingParams::FromQuality(quality_);
+    R.Initialize(params.lobes, params.lookup_size);
+  }
+
+  void RunImpl(HostWorkspace &ws) override {
+    auto &out = ws.template Output<CPUBackend>(0);
+    const auto &in = ws.template Input<CPUBackend>(0);
+
+    const auto &out_shape = out.shape();
+    const auto &in_shape = in.shape();
+    out.SetLayout(in.GetLayout());
+    int N = in.num_samples();
+    assert(N == static_cast<int>(scales_.size()));
+    assert(out.type() == dtype_);
+
+    auto &tp = ws.GetThreadPool();
+    in_fp32.resize(tp.NumThreads());
+    for (int s = 0; s < N; s++) {
+      tp.AddWork([&, this, s](int thread_idx) {
+        InTensorCPU<float> in_view;
+        TYPE_SWITCH(in.type(), type2id, T, (AUDIO_RESAMPLE_TYPES),
+          (in_view = ConvertInput(in_fp32[thread_idx], view<const T>(in[s]));),
+          (DALI_FAIL(
+              make_string("Unsupported output type: ", dtype_,
+                          "\nSupported types are : ", ListTypeNames<AUDIO_RESAMPLE_TYPES>()));));
+        TYPE_SWITCH(dtype_, type2id, T, (AUDIO_RESAMPLE_TYPES),
+          (ResampleTyped<T>(view<T>(out[s]), in_view, scales_[s]);),
+          (assert(!"Unreachable code.")));
+      });
+    }
+    tp.RunAll();
+  }
+
+  template <typename T>
+  void ResampleTyped(const OutTensorCPU<T> &out, const InTensorCPU<float> &in, double scale) {
+    int ch = out.shape.sample_dim() > 1 ? out.shape[1] : 1;
+    R.Resample(out.data, 0, out.shape[0], scale, in.data, in.shape[0], 1, ch);
+  }
+
+  template <typename T>
+  InTensorCPU<float> ConvertInput(std::vector<float> &tmp, const InTensorCPU<T> &in) {
```

Comment:
I'm not sure if `ConvertInput` needs to be a member. All it uses is just `dtype_`, other parameters are passed as arguments.